File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0
        res = [0]
        def dfs(i, j, n, m,res):
            if (i < 0 or i >= m) or (j < 0 or j >=n) or grid[i][j] == 0:
                res[0] += 1
                return
            if grid[i][j] == -1:
                return
            grid[i][j] = -1
            dfs(i+1, j, n, m,res)
            dfs(i, j+1, n, m,res)
            dfs(i-1, j, n, m,res)
            dfs(i, j-1, n, m,res)
        
        m = len(grid)
        n = len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    dfs(i, j,n, m,res)
        return res[0]

File: 0463-island-perimeter/0463-island-perimeter.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& grid,vector<vector<int>> &vis,int i, int j, int m, int n, int &ans){
       if( i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == 0){
           ans++;
           return;
       }
       if(vis[i][j] == 1){
           return;
       }
       vis[i][j] = 1;
        dfs(grid, vis, i + 1, j, m, n, ans);
        dfs(grid, vis, i - 1, j, m, n,ans);
        dfs(grid, vis, i, j + 1, m, n, ans);
        dfs(grid, vis, i, j - 1, m, n, ans);
    }
    int islandPerimeter(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int ans = 0;
        vector<vector<int>> vis(m, vector<int>(n,0));
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if( grid[i][j] == 1){
                    dfs(grid,vis, i, j, m, n, ans);
                }
            }
        }
        return ans;
    }
};